.changeset/fruity-birds-love.md

```diff
@@ -0,0 +1,34 @@
+---
+"fluid-framework": minor
+"@fluidframework/tree": minor
+---
+---
+"section": "tree"
+---
+Enable compile time type narrowing based on a TreeNode's NodeKind.
+
+TreeNode's schema aware APIs implement WithType, which now has a NodeKind parameter that can be used to narrow TreeNodes based on NodeKind.
+
+Example:
+
+```typescript
+function getKeys(node: TreeNode & WithType<string, NodeKind.Array>): number[];
+function getKeys(node: TreeNode & WithType<string, NodeKind.Map | NodeKind.Object>): string[];
+function getKeys(node: TreeNode): string[] | number[];
+function getKeys(node: TreeNode): string[] | number[] {
+	const schema = Tree.schema(node);
```

Comment:
Best practices question, since this example is end-user-facing: should we use a `switch` here with a default handler that errors on unrecognized node kind? I don't know how likely it is that we will add a fundamentally new node type any time soon, but if we were to do so, the default behavior here may not be what the user wants.

Seems like the switch pattern highlights this potential better. Curious what others think.

Reply:
Done.